org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs

```diff
@@ -16,63 +16,99 @@ namespace MixedReality.Toolkit.Input
+
+            foreach (IXRProximityInteractable previouslyDetectedInteractable in previouslyDetectedColliderInteractableDuples)
+            {
+                if (!currentlyDetectedInteractable.Contains(previouslyDetectedInteractable))
+                {
+                    noLongerDetectedInteractables.Add(previouslyDetectedInteractable);
+                }
+            }
+
+            foreach (IXRProximityInteractable noLongerDetectedInteractable in noLongerDetectedInteractables)
+            {
+                noLongerDetectedInteractable.OnProximityExited(new ProximityHoverExitedEventArgs(noLongerDetectedInteractable));
+                currentlyDetectedInteractable.Remove(noLongerDetectedInteractable);
+            }
+        }
+
+        /// <summary>
+        /// Call OnProximityEntered for all colliders that are currently detected but were not detected previously.
+        /// </summary>
+        private void UpdateProximityEntered()
+        {
+            List<IXRProximityInteractable> currentlyDetectedInteractables = GetCurrentlyDetectedColliderInteractablesDuples();
+            foreach (IXRProximityInteractable colliderInteractableDuple in currentlyDetectedInteractables)
+            {
+                if (previouslyDetectedColliderInteractableDuples.Add(colliderInteractableDuple))
                 {
-                    Collider previouslyDetectedCollider = previouslyDetectedColliders[i];
-                    if (!DetectedColliders.Contains(previouslyDetectedCollider) && previouslyDetectedCollider != null)
-                    {
-                        IXRProximityInteractable nearInteractionMode = previouslyDetectedCollider.GetComponent<IXRProximityInteractable>();
-                        if (nearInteractionMode != null)
-                        {
-                            foreach (XRBaseInteractor xrBaseInteractor in nearInteractors)
-                            {
-                                previouslyDetectedCollider.GetComponent<IXRProximityInteractable>().OnProximityExited(new ProximityExitedEventArgs(this, previouslyDetectedCollider, xrBaseInteractor));
-                            }
-                        }
-                        previouslyDetectedColliders.Remove(previouslyDetectedCollider);
-                    }
+                    colliderInteractableDuple.OnProximityEntered(new ProximityHoverEnteredEventArgs(colliderInteractableDuple));
                 }
-                foreach (Collider collider in DetectedColliders)
+            }
+        }
+
+        /// <summary>
+        /// Returns a hashset of all unique interactables that are currently detected.
+        /// </summary>
+        /// <returns>Hashset with unique interactables currently detected</returns>
+        private List<IXRProximityInteractable> GetCurrentlyDetectedColliderInteractablesDuples()
+        {
+            List<IXRProximityInteractable> result = new();
```

Comment:
Remove "per frame" allocation.   Instead have class member list that you only allocate once, and clear() every frame.

Reply:
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/659/commits/7ee0fe33c7946426a3a643e4c266e4defd1fc3e2 resolving